---
title: "Tutorial: Make the Best of Bioconductor Workflows"
output:
  BiocStyle::html_document:
    toc: true
    number_sections: true
    highlight: haddock
    css: style.css
---

<!--
%\VignetteIndexEntry{Tutorial: make the best of bioconductor workflows}
%\VignettePackage{sevenbridges}
%\VignetteEngine{knitr::rmarkdown}
-->

```{r style, echo = FALSE, results = 'asis'}
BiocStyle::markdown(css.files = "custom.css")
```

<!-- override white background for classless code chunks -->
<style type="text/css" scoped>
  pre:not([class]) { background-color: #F8F8F8; }
  pre code { background-color: transparent; }
</style>

```{r code, echo = FALSE}
code <- function(...) {
    cat(paste(..., sep = "\n"))
}

code2 <- function(...) {
    cat(paste("```markdown", ..., "\n", "```", sep = "\n"))
}
```



# Introduction

__In Progess!!!!__

__Goal__: I hope to encourage Biocondcutor developer and R developers to find a formal way to make their tool/flow into a reproducible container with standard command line interface or provide official "executable" report template with knitr. (Hope to have more fun with liftr soon later).

In this tutorial, I will show you 

1. Wrap existing bioconductor workflow into a container
2. Provide command line interface for your container app
3. Provide template report for your container app
4. Run it anywhere include
    - Locally
    - Cancer genomics cloud
    - Google 
    - AWS 
5. Check they are the same

# RNA-seq analysis pipeline




## Creat a Tool: from Bam to report

### Step 1: have a plan 
To demonstrate I will use a RNA-seq [workfow](https://www.bioconductor.org/help/workflows/rnaseqGene/) from Bioconductor, I want to finish two tasks

- Input raw bam and design matrics
- Give two outputs
    - Report
    - Graphics
    - DE table



### Step 2: Create docker container for your tool

Why we do this first? Building a development environment is essential for developing your flow command line interface, so we should get everything prepared. Saved thousands words here, please check what do in my Dockerfile, 
I installed the official "workflow" first from Bioconducotr. 

```{r}
fl <- system.file("docker/rnaseqgene", "Dockerfile", 
                  package = "sevenbridges")
fl
```

Here is the current content of Dockerfile
```{r comment=''}
cat(readLines(fl), sep = '\n')
```


### Step 3: Create your command line interface.

In this step, I am going to

1. Use docopt to make R command line interface
2. Insert my script into the container I made
3. Test function inside container or
4. Test function locally using rabix

### Step 4: Describe your tool in R into CWL.

```{r}
library(sevenbridges)

rbx <- Tool(id = "rnaseqGene", 
            label = "rnaseqgene",
            description = "A RNA-seq Differiencial Expression Flow and Report",
            hints = requirements(docker(pull = "tengfei/rnaseqgene"), cpu(1), mem(2000)), 
            baseCommand = "performDE.R", 
            inputs = list(
                input(
                    id = "bamfiles", label = "bam files",
                    description = "a list of bam files",
                    type = ItemArray("File"),
                    prefix = "--bamfiles",
                    itemSeparator = ","
                ), 
                input(
                    id = "design", label = "design matrix",
                    type = "File",
                    prefix = "--design"
                ),
                input(
                    id = "gtffile", label =  "gene feature files",
                    type = "File",
                    prefix = "--gtffile"
                )
            ),
            outputs = list(
                output(id = "report", label = "report", 
                       description = "A reproducible report created by Rmarkdown",
                       glob = "*.pdf"),
                output(id = "heatmap", label = "heatmap", 
                       description = "A heatmap plot to show the Euclidean distance between samples",
                       glob = "*.pdf"),
                output(id = "count", label = "count", 
                       description = "Reads counts matrix",
                       glob = "*.csv")))


rbx$toJSON()

rbx$toJSON(pretty = TRUE)

## or write to external file
fl <- "~/Downloads/rnaseqGene.json"
write(rbx$toJSON(), fl)
file.show(fl)
```

Now let's create a task and execute it with example files. 

```{r}
library(sevenbridges)
## load from preset authentification
a <- Auth(username = "tengfei", platform = "us")
a
## Load the project
(p <- a$project(id = "tengfei/rnaseqgene"))

## get files you need as inout
(bamfiles.in <- p$file(".bam"))
(design.in <- p$file("sample_table.csv"))
(gtf.in <- p$file("Homo_sapiens.GRCh37.75_subset.gtf"))

## add App you just created 
(rna.app <- p$app_add("rnaseqgene7", rbx))

## add a new Task
(tsk <- p$task_add(name = "RNA DE report 3", 
                   description = "RNA DE analysis report", 
                   app = rna.app$id,
                   inputs = list(bamfiles = bamfiles.in, 
                                 design = design.in,
                                 gtffile = gtf.in)))

## run the task
tsk$run()

## monitor the task
tsk$monitor()
```

```
rabix ~/Code/svnrepos/bioc-devel/sevenbridges/inst/docker/rnaseqGene/rabix/rnaseqGene.json -- --bamfiles SRR1039508_subset.bam,SRR1039509_subset.bam,SRR1039512_subset.bam,SRR1039513_subset.bam,SRR1039516_subset.bam,SRR1039517_subset.bam,SRR1039520_subset.bam,SRR1039521_subset.bam --design sample_table.csv --gtffile Homo_sapiens.GRCh37.75_subset.gtf
```

### Step 5: Execute it.

Please read next section about execution. 

## Execute it anywhere

### Create a project

I am going to create a project that host my apps and test data.

```{r, eval = FALSE}
a <- Auth(platform = "cgc", username = "tengfei")
bid <- a$billing()[[1]]$id
a$project_new("Bioconductor RNA-seq Workflow", billing_group_id = bid,
              description = "An bam to report workflow for RNA-seq analysis")

```

### Upload your test data



### Push your apps 

### Run it!



### Cancer Genomics Cloud

### Google cloud and other seven bridges platforms



## Connecting with existing App  

Don't always reinvent the wheel, unless you can produce something better. 

For example, now I want to create a workflow that use STAR aligner to do the aignments first. Then I searched on line, I found "oh yes, there is already an app on Cancer Genomics Cloud public app repos."

## Share your tools


    
