<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Seven Bridges Platform API Client and Common Workflow Language Tool
    Builder in R • sevenbridges-r</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><link href="extra.css" rel="stylesheet">
<meta property="og:title" content="Seven Bridges Platform API Client and Common Workflow Language Tool
    Builder in R">
<meta property="og:description" content="R client and utilities for Seven Bridges platform API, from Cancer
    Genomics Cloud to other Seven Bridges supported platforms.">
<meta property="og:image" content="/logo.png">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-87185344-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-87185344-1');
</script>
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">sevenbridges-r</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.15.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/api.html">Complete Guide for Seven Bridges API R Client</a>
    </li>
    <li>
      <a href="articles/apps.html">Describe and Execute CWL Tools/Workflows in R</a>
    </li>
    <li>
      <a href="articles/bioc-workflow.html">Master Tutorial: Use R for Cancer Genomics Cloud</a>
    </li>
    <li>
      <a href="articles/cgc-datasets.html">Find Data on CGC via Data Browser and Datasets API</a>
    </li>
    <li>
      <a href="articles/docker.html">Creating Your Docker Container and Command Line Interface (with docopt)</a>
    </li>
    <li>
      <a href="articles/rstudio.html">IDE Container: RStudio Server, Shiny Server, and More</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/sbg/sevenbridges-r">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="contents col-md-9">
<div id="sevenbridges-r-" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#sevenbridges-r-" class="anchor"></a>sevenbridges-r <a href="https://www.sevenbridges.com"><img src="logo.png" align="right" alt="logo" height="128" width="128"></a>
</h1></div>

<p><a href="https://www.bioconductor.org/packages/release/bioc/html/sevenbridges.html">BioC (Release)</a> · <a href="https://www.bioconductor.org/packages/devel/bioc/html/sevenbridges.html">BioC (Development)</a> · <a href="https://github.com/sbg/sevenbridges-r">GitHub (Latest)</a></p>
<div id="overview" class="section level2">
<h2 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h2>
<p>sevenbridges-r is an <a href="https://www.r-project.org/">R</a>/<a href="https://www.bioconductor.org/">Bioconductor</a> package that provides an interface for the <a href="https://www.sevenbridges.com/">Seven Bridges Platform</a> (US, EU, China), <a href="https://www.cancergenomicscloud.org/">Cancer Genomics Cloud</a>, <a href="http://www.cavatica.org/">Cavatica</a>, and <a href="https://f4c.sbgenomics.com">DataSTAGE Powered by Seven Bridges</a> public APIs.</p>
<p>The <a href="https://www.sevenbridges.com/">Seven Bridges Platform</a> is a cloud-based environment for conducting bioinformatics analysis. It is a central hub for teams to store, analyze, and jointly interpret their bioinformatic data. The Platform co-locates analysis pipelines alongside the largest genomic datasets to optimize processing, allocating storage, and compute resources on demand.</p>
<p>The <a href="https://www.cancergenomicscloud.org/">Cancer Genomics Cloud (CGC)</a>, powered by <a href="https://www.sevenbridges.com/">Seven Bridges</a>, is also a cloud-based computation environment. It was built as one of three pilot systems funded by the <a href="https://www.cancer.gov">National Cancer Institute</a> to explore the paradigm of colocalizing massive genomics datasets, like The <a href="https://cancergenome.nih.gov">Cancer Genomics Atlas (TCGA)</a>, alongside secure and scalable computational resources to analyze them. The CGC makes more than a petabyte of multi-dimensional data available immediately to authorized researchers. You can add your data to analyze alongside TCGA using predefined analytical workflows or your own tools.</p>
<p><a href="http://www.cavatica.org/">Cavatica</a>, powered by <a href="https://www.sevenbridges.com">Seven Bridges</a>, is a data analysis and sharing platform designed to accelerate discovery in a scalable, cloud-based compute environment where data, results, and workflows are shared among the world’s research community. Cavatica is built in collaboration with the Children Hospital of Philadelphia and it is focused on pediatric data.</p>
</div>
<div id="table-of-contents" class="section level2">
<h2 class="hasAnchor">
<a href="#table-of-contents" class="anchor"></a>Table of Contents</h2>
<ul>
<li><a href="#overview">Overview</a></li>
<li>
<a href="#installation">Installation</a>
<ul>
<li><a href="#check-r-version">Check R version</a></li>
<li><a href="#bioconductor---release-branch">Bioconductor - Release Branch</a></li>
<li><a href="#bioconductor---development-branch">Bioconductor - Development Branch</a></li>
<li><a href="#latest-development-version">Latest Development Version</a></li>
</ul>
</li>
<li>
<a href="#features">Features</a>
<ul>
<li><a href="#flexible-authentication-methods">Flexible Authentication Methods</a></li>
<li><a href="#complete-api-r-client">Complete API R Client</a></li>
<li><a href="#task-monitoring">Task Monitoring</a></li>
<li><a href="#batch-tasks-support">Batch Tasks Support</a></li>
<li><a href="#cross-environment-support">Cross Environment Support</a></li>
<li><a href="#common-workflow-language-tool-interface">Common Workflow Language Tool Interface</a></li>
<li><a href="#utilities-for-tool-and-flow">Utilities for Tool and Flow</a></li>
</ul>
</li>
<li><a href="#tutorials">Tutorials</a></li>
<li><a href="#ide-docker-image">IDE Docker Image</a></li>
<li><a href="#faq">FAQ</a></li>
<li><a href="#events">Events</a></li>
<li><a href="#contribute">Contribute</a></li>
<li><a href="#copyright">Copyright</a></li>
</ul>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<div id="check-r-version" class="section level3">
<h3 class="hasAnchor">
<a href="#check-r-version" class="anchor"></a>Check R version</h3>
<p>First, check the version of R you are using with the following command (in R):</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a>R.version.string</span></code></pre></div>
<p>If you are not running the latest release version of R, install or upgrade with <a href="https://www.bioconductor.org/install/#install-R">these instructions</a>. If you are using RStudio, restart RStudio after installing R. RStudio will detect the new installation.</p>
</div>
<div id="bioconductor---release-branch" class="section level3">
<h3 class="hasAnchor">
<a href="#bioconductor---release-branch" class="anchor"></a>Bioconductor - Release Branch</h3>
<p>This is recommended for most users as it is the most stable version.</p>
<p>You can install the package from the <code>release</code> branch on Bioconductor using <code>BiocManager</code>:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages</a></span>(<span class="st">"BiocManager"</span>)</span>
<span id="cb2-2"><a href="#cb2-2"></a>BiocManager<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/BiocManager/topics/install">install</a></span>(<span class="st">"sevenbridges"</span>)</span></code></pre></div>
</div>
<div id="bioconductor---development-branch" class="section level3">
<h3 class="hasAnchor">
<a href="#bioconductor---development-branch" class="anchor"></a>Bioconductor - Development Branch</h3>
<p>If you are developing tools under the <code>devel</code> branch or use the development version of R and Bioconductor, install the package from the Bioconductor <code>devel</code> branch. You probably also want to install R-devel first by following the directions in <a href="https://bioconductor.org/developers/how-to/useDevel/">“Using the ‘Devel’ Version of Bioconductor”</a>.</p>
<p>To install the <code>sevenbridges</code> package from the <code>devel</code> branch, use</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages</a></span>(<span class="st">"BiocManager"</span>)</span>
<span id="cb3-2"><a href="#cb3-2"></a>BiocManager<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/BiocManager/topics/install">install</a></span>(<span class="st">"sevenbridges"</span>, <span class="dt">version =</span> <span class="st">"devel"</span>)</span></code></pre></div>
</div>
<div id="latest-development-version" class="section level3">
<h3 class="hasAnchor">
<a href="#latest-development-version" class="anchor"></a>Latest Development Version</h3>
<p>To try the latest features, please install the package directly from GitHub. We push to the Bioconductor branch (<code>release</code> and <code>devel</code>) regularly.</p>
<p>Installing the <code>sevenbridges</code> package from GitHub requires you have the <code>devtools</code> package. If you do not have <code>devtools</code>, install it from CRAN first.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages</a></span>(<span class="st">"devtools"</span>)</span></code></pre></div>
<p>You may get an error for missing system dependecies such as curl and ssl. You probably need to do the following first in order to install <code>devtools</code> and to build vignettes since you need <code>pandoc</code> under Ubuntu.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1"></a><span class="ex">apt-get</span> update</span>
<span id="cb5-2"><a href="#cb5-2"></a><span class="ex">apt-get</span> install libcurl4-gnutls-dev libssl-dev pandoc pandoc-citeproc</span></code></pre></div>
<p>After <code>devtools</code> is installed, install the latest version of <code>sevenbridges</code> from GitHub:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages</a></span>(<span class="st">"BiocManager"</span>)</span>
<span id="cb6-2"><a href="#cb6-2"></a></span>
<span id="cb6-3"><a href="#cb6-3"></a>devtools<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/devtools/topics/reexports">install_github</a></span>(</span>
<span id="cb6-4"><a href="#cb6-4"></a>  <span class="st">"sbg/sevenbridges-r"</span>,</span>
<span id="cb6-5"><a href="#cb6-5"></a>  <span class="dt">repos =</span> BiocManager<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/BiocManager/topics/repositories">repositories</a></span>(),</span>
<span id="cb6-6"><a href="#cb6-6"></a>  <span class="dt">build_vignettes =</span> <span class="ot">TRUE</span>, <span class="dt">dependencies =</span> <span class="ot">TRUE</span></span>
<span id="cb6-7"><a href="#cb6-7"></a>)</span></code></pre></div>
<p>If you have trouble with <code>pandoc</code> and do not want to install it, set <code>build_vignettes = FALSE</code> to avoid building the vignettes.</p>
</div>
</div>
<div id="features" class="section level2">
<h2 class="hasAnchor">
<a href="#features" class="anchor"></a>Features</h2>
<p>The <code>sevenbridges</code> package includes the following features:</p>
<div id="flexible-authentication-methods" class="section level3">
<h3 class="hasAnchor">
<a href="#flexible-authentication-methods" class="anchor"></a>Flexible Authentication Methods</h3>
<p>Multiple authentication methods support.</p>
<ul>
<li>Direct authentication:</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="co"># Direct authentication</span></span>
<span id="cb7-2"><a href="#cb7-2"></a>a &lt;-<span class="st"> </span><span class="kw"><a href="reference/Auth-class.html">Auth</a></span>(<span class="dt">token =</span> <span class="st">"your_token"</span>, <span class="dt">platform =</span> <span class="st">"cgc"</span>)</span>
<span id="cb7-3"><a href="#cb7-3"></a></span>
<span id="cb7-4"><a href="#cb7-4"></a><span class="co"># or use base url</span></span>
<span id="cb7-5"><a href="#cb7-5"></a>a &lt;-<span class="st"> </span><span class="kw"><a href="reference/Auth-class.html">Auth</a></span>(<span class="dt">token =</span> <span class="st">"your_token"</span>, <span class="dt">url =</span> <span class="st">"https://cgc-api.sbgenomics.com/v2"</span>)</span></code></pre></div>
<ul>
<li>Authentication via system environment variables:</li>
</ul>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="kw"><a href="reference/sbg_set_env.html">sbg_set_env</a></span>(<span class="dt">token =</span> <span class="st">"your_token"</span>, <span class="dt">url =</span> <span class="st">"https://cgc-api.sbgenomics.com/v2"</span>)</span>
<span id="cb8-2"><a href="#cb8-2"></a>a &lt;-<span class="st"> </span><span class="kw"><a href="reference/Auth-class.html">Auth</a></span>(<span class="dt">from =</span> <span class="st">"env"</span>)</span></code></pre></div>
<ul>
<li>Authentication via a user configuration file, collect and manage your credentials for multiple accounts across various Seven Bridges environments:</li>
</ul>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a>a &lt;-<span class="st"> </span><span class="kw"><a href="reference/Auth-class.html">Auth</a></span>(<span class="dt">from =</span> <span class="st">"file"</span>, <span class="dt">profile_name =</span> <span class="st">"aws-us-username"</span>)</span></code></pre></div>
<p>Please check <code><a href="articles/api.html">vignette("api", package = "sevenbridges")</a></code> for technical details about all available authentication methods.</p>
</div>
<div id="complete-api-r-client" class="section level3">
<h3 class="hasAnchor">
<a href="#complete-api-r-client" class="anchor"></a>Complete API R Client</h3>
<p>A complete API R client with a user-friendly, object-oriented API with printing and support operations for API requests relating to users, billing, projects, files, apps, and tasks. Short examples are also included, as shown below:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a><span class="co"># Get a project by pattern-matching its name</span></span>
<span id="cb10-2"><a href="#cb10-2"></a>p &lt;-<span class="st"> </span>a<span class="op">$</span><span class="kw">project</span>(<span class="st">"demo"</span>)</span>
<span id="cb10-3"><a href="#cb10-3"></a></span>
<span id="cb10-4"><a href="#cb10-4"></a><span class="co"># Get a project by its id</span></span>
<span id="cb10-5"><a href="#cb10-5"></a>p &lt;-<span class="st"> </span>a<span class="op">$</span><span class="kw">project</span>(<span class="dt">id =</span> <span class="st">"username/demo"</span>)</span>
<span id="cb10-6"><a href="#cb10-6"></a></span>
<span id="cb10-7"><a href="#cb10-7"></a><span class="co"># Delete files from a project</span></span>
<span id="cb10-8"><a href="#cb10-8"></a>p<span class="op">$</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/connections">file</a></span>(<span class="st">"sample.tz"</span>)<span class="op">$</span><span class="kw"><a href="reference/delete-methods.html">delete</a></span>()</span>
<span id="cb10-9"><a href="#cb10-9"></a></span>
<span id="cb10-10"><a href="#cb10-10"></a><span class="co"># Upload fies from a folder to a project and include file metadata</span></span>
<span id="cb10-11"><a href="#cb10-11"></a>p<span class="op">$</span><span class="kw">upload</span>(<span class="st">"folder_path"</span>, <span class="dt">metadata =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">platform =</span> <span class="st">"Illumina"</span>))</span></code></pre></div>
</div>
<div id="task-monitoring" class="section level3">
<h3 class="hasAnchor">
<a href="#task-monitoring" class="anchor"></a>Task Monitoring</h3>
<p>A task monitoring hook which allows you to add a hook function to specific task statuses as you monitor a task. For example, you can opt to receive an email when the task is completed or specify to download all files produced by the task, as shown below:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a><span class="kw"><a href="reference/TaskHook.html">setTaskHook</a></span>(<span class="st">"completed"</span>, <span class="cf">function</span>() {</span>
<span id="cb11-2"><a href="#cb11-2"></a>  tsk<span class="op">$</span><span class="kw"><a href="reference/download-methods.html">download</a></span>(<span class="st">"~/Downloads"</span>)</span>
<span id="cb11-3"><a href="#cb11-3"></a>})</span>
<span id="cb11-4"><a href="#cb11-4"></a>tsk<span class="op">$</span><span class="kw">monitor</span>()</span></code></pre></div>
</div>
<div id="batch-tasks-support" class="section level3">
<h3 class="hasAnchor">
<a href="#batch-tasks-support" class="anchor"></a>Batch Tasks Support</h3>
<p>Batch tasks by metadata and by item.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a><span class="co"># Batch by item</span></span>
<span id="cb12-2"><a href="#cb12-2"></a>(tsk &lt;-<span class="st"> </span>p<span class="op">$</span><span class="kw">task_add</span>(</span>
<span id="cb12-3"><a href="#cb12-3"></a>  <span class="dt">name =</span> <span class="st">"RNA DE report new batch 2"</span>,</span>
<span id="cb12-4"><a href="#cb12-4"></a>  <span class="dt">description =</span> <span class="st">"RNA DE analysis report"</span>,</span>
<span id="cb12-5"><a href="#cb12-5"></a>  <span class="dt">app =</span> rna.app<span class="op">$</span>id,</span>
<span id="cb12-6"><a href="#cb12-6"></a>  <span class="dt">batch =</span> <span class="kw"><a href="reference/batch.html">batch</a></span>(<span class="dt">input =</span> <span class="st">"bamfiles"</span>),</span>
<span id="cb12-7"><a href="#cb12-7"></a>  <span class="dt">inputs =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(</span>
<span id="cb12-8"><a href="#cb12-8"></a>    <span class="dt">bamfiles =</span> bamfiles.in,</span>
<span id="cb12-9"><a href="#cb12-9"></a>    <span class="dt">design =</span> design.in,</span>
<span id="cb12-10"><a href="#cb12-10"></a>    <span class="dt">gtffile =</span> gtf.in</span>
<span id="cb12-11"><a href="#cb12-11"></a>  )</span>
<span id="cb12-12"><a href="#cb12-12"></a>))</span>
<span id="cb12-13"><a href="#cb12-13"></a></span>
<span id="cb12-14"><a href="#cb12-14"></a><span class="co"># Batch by metadata. Note that input files must</span></span>
<span id="cb12-15"><a href="#cb12-15"></a><span class="co"># have relevant metadata fields specified.</span></span>
<span id="cb12-16"><a href="#cb12-16"></a>(tsk &lt;-<span class="st"> </span>p<span class="op">$</span><span class="kw">task_add</span>(</span>
<span id="cb12-17"><a href="#cb12-17"></a>  <span class="dt">name =</span> <span class="st">"RNA DE report new batch 3"</span>,</span>
<span id="cb12-18"><a href="#cb12-18"></a>  <span class="dt">description =</span> <span class="st">"RNA DE analysis report"</span>,</span>
<span id="cb12-19"><a href="#cb12-19"></a>  <span class="dt">app =</span> rna.app<span class="op">$</span>id,</span>
<span id="cb12-20"><a href="#cb12-20"></a>  <span class="dt">batch =</span> <span class="kw"><a href="reference/batch.html">batch</a></span>(</span>
<span id="cb12-21"><a href="#cb12-21"></a>    <span class="dt">input =</span> <span class="st">"fastq"</span>,</span>
<span id="cb12-22"><a href="#cb12-22"></a>    <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"metadata.sample_id"</span>, <span class="st">"metadata.library_id"</span>)</span>
<span id="cb12-23"><a href="#cb12-23"></a>  ),</span>
<span id="cb12-24"><a href="#cb12-24"></a>  <span class="dt">inputs =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(</span>
<span id="cb12-25"><a href="#cb12-25"></a>    <span class="dt">bamfiles =</span> bamfiles.in,</span>
<span id="cb12-26"><a href="#cb12-26"></a>    <span class="dt">design =</span> design.in,</span>
<span id="cb12-27"><a href="#cb12-27"></a>    <span class="dt">gtffile =</span> gtf.in</span>
<span id="cb12-28"><a href="#cb12-28"></a>  )</span>
<span id="cb12-29"><a href="#cb12-29"></a>))</span></code></pre></div>
</div>
<div id="cross-environment-support" class="section level3">
<h3 class="hasAnchor">
<a href="#cross-environment-support" class="anchor"></a>Cross Environment Support</h3>
<p>Cross-platform support for Seven Bridges environments, such as <a href="https://www.cancergenomicscloud.org/">Cancer Genomics Cloud</a> or <a href="https://www.sbgenomics.com/">Seven Bridges Platform</a> on either Amazon Web Services or Google Cloud Platform.</p>
</div>
<div id="common-workflow-language-tool-interface" class="section level3">
<h3 class="hasAnchor">
<a href="#common-workflow-language-tool-interface" class="anchor"></a>Common Workflow Language Tool Interface</h3>
<p>A <a href="http://www.commonwl.org/">Common Workflow Language (CWL)</a> Tool interface to directly describe your tool in R, export it to JSON or YAML, or add it to your online project. This package defines a complete set of CWL object, so you can describe tools as follows:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a>fd &lt;-<span class="st"> </span><span class="kw"><a href="reference/requirements.html">fileDef</a></span>(<span class="dt">name =</span> <span class="st">"runif.R"</span>, <span class="dt">content =</span> readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/read_file.html">read_file</a></span>(fl))</span>
<span id="cb13-2"><a href="#cb13-2"></a></span>
<span id="cb13-3"><a href="#cb13-3"></a>rbx &lt;-<span class="st"> </span><span class="kw"><a href="reference/Tool-class.html">Tool</a></span>(</span>
<span id="cb13-4"><a href="#cb13-4"></a>  <span class="dt">id =</span> <span class="st">"runif"</span>,</span>
<span id="cb13-5"><a href="#cb13-5"></a>  <span class="dt">label =</span> <span class="st">"runif"</span>,</span>
<span id="cb13-6"><a href="#cb13-6"></a>  <span class="dt">hints =</span> <span class="kw"><a href="reference/requirements.html">requirements</a></span>(</span>
<span id="cb13-7"><a href="#cb13-7"></a>    <span class="kw"><a href="reference/requirements.html">docker</a></span>(<span class="dt">pull =</span> <span class="st">"rocker/r-base"</span>),</span>
<span id="cb13-8"><a href="#cb13-8"></a>    <span class="kw"><a href="reference/requirements.html">cpu</a></span>(<span class="dv">1</span>), <span class="kw"><a href="reference/requirements.html">mem</a></span>(<span class="dv">2000</span>)</span>
<span id="cb13-9"><a href="#cb13-9"></a>  ),</span>
<span id="cb13-10"><a href="#cb13-10"></a>  <span class="dt">requirements =</span> <span class="kw"><a href="reference/requirements.html">requirements</a></span>(fd),</span>
<span id="cb13-11"><a href="#cb13-11"></a>  <span class="dt">baseCommand =</span> <span class="st">"Rscript runif.R"</span>,</span>
<span id="cb13-12"><a href="#cb13-12"></a>  <span class="dt">stdout =</span> <span class="st">"output.txt"</span>,</span>
<span id="cb13-13"><a href="#cb13-13"></a>  <span class="dt">inputs =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(</span>
<span id="cb13-14"><a href="#cb13-14"></a>    <span class="kw"><a href="reference/shorthand-cwl.html">input</a></span>(<span class="dt">id =</span> <span class="st">"number"</span>, <span class="dt">type =</span> <span class="st">"integer"</span>, <span class="dt">position =</span> <span class="dv">1</span>),</span>
<span id="cb13-15"><a href="#cb13-15"></a>    <span class="kw"><a href="reference/shorthand-cwl.html">input</a></span>(<span class="dt">id =</span> <span class="st">"min"</span>, <span class="dt">type =</span> <span class="st">"float"</span>, <span class="dt">position =</span> <span class="dv">2</span>),</span>
<span id="cb13-16"><a href="#cb13-16"></a>    <span class="kw"><a href="reference/shorthand-cwl.html">input</a></span>(<span class="dt">id =</span> <span class="st">"max"</span>, <span class="dt">type =</span> <span class="st">"float"</span>, <span class="dt">position =</span> <span class="dv">3</span>)</span>
<span id="cb13-17"><a href="#cb13-17"></a>  ),</span>
<span id="cb13-18"><a href="#cb13-18"></a>  <span class="dt">outputs =</span> <span class="kw"><a href="reference/shorthand-cwl.html">output</a></span>(<span class="dt">id =</span> <span class="st">"random"</span>, <span class="dt">glob =</span> <span class="st">"output.txt"</span>)</span>
<span id="cb13-19"><a href="#cb13-19"></a>)</span>
<span id="cb13-20"><a href="#cb13-20"></a></span>
<span id="cb13-21"><a href="#cb13-21"></a><span class="co"># Print CWL JSON</span></span>
<span id="cb13-22"><a href="#cb13-22"></a>rbx<span class="op">$</span><span class="kw">toJSON</span>(<span class="dt">pretty =</span> <span class="ot">TRUE</span>)</span>
<span id="cb13-23"><a href="#cb13-23"></a></span>
<span id="cb13-24"><a href="#cb13-24"></a><span class="co"># Print CWL YAML</span></span>
<span id="cb13-25"><a href="#cb13-25"></a>rbx<span class="op">$</span><span class="kw">toYAML</span>()</span></code></pre></div>
</div>
<div id="utilities-for-tool-and-flow" class="section level3">
<h3 class="hasAnchor">
<a href="#utilities-for-tool-and-flow" class="anchor"></a>Utilities for Tool and Flow</h3>
<p>Utilities for <code>Tool</code> and <code>Flow</code>, for example</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(<span class="st">"sevenbridges"</span>)</span>
<span id="cb14-2"><a href="#cb14-2"></a></span>
<span id="cb14-3"><a href="#cb14-3"></a><span class="co"># convert a SBG CWL JSON file</span></span>
<span id="cb14-4"><a href="#cb14-4"></a>t1 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata/app"</span>, <span class="st">"tool_star.json"</span>, <span class="dt">package =</span> <span class="st">"sevenbridges"</span>)</span>
<span id="cb14-5"><a href="#cb14-5"></a></span>
<span id="cb14-6"><a href="#cb14-6"></a><span class="co"># convert json file into a Tool object</span></span>
<span id="cb14-7"><a href="#cb14-7"></a>t1 &lt;-<span class="st"> </span><span class="kw"><a href="reference/convert_app.html">convert_app</a></span>(t1)</span>
<span id="cb14-8"><a href="#cb14-8"></a></span>
<span id="cb14-9"><a href="#cb14-9"></a><span class="co"># shows all input matrix</span></span>
<span id="cb14-10"><a href="#cb14-10"></a>t1<span class="op">$</span><span class="kw"><a href="reference/input_output_matrix.html">input_matrix</a></span>()</span></code></pre></div>
</div>
</div>
<div id="tutorials" class="section level2">
<h2 class="hasAnchor">
<a href="#tutorials" class="anchor"></a>Tutorials</h2>
<p>We maintain 3 different sets of documentation: the sevenbridges-r GitHub repository (latest and most up-to-date), Bioconductor release channel, and Bioconductor development channel. Below, only the GitHub version is linked to provide the latest documentation. For the other versions, please visit <a href="https://www.bioconductor.org/packages/release/bioc/html/sevenbridges.html">Bioconductor Release version</a> or <a href="https://www.bioconductor.org/packages/devel/bioc/html/sevenbridges.html">Bioconductor Development version</a>. The tutorials below are re-generated regularly as we update the package on GitHub.</p>
<table class="table">
<colgroup>
<col width="66%">
<col width="15%">
<col width="18%">
</colgroup>
<thead><tr class="header">
<th align="left">Tutorial Title</th>
<th align="left">HTML</th>
<th align="left">Rmd Source</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Complete Reference for the API R Client</td>
<td align="left"><a href="https://sbg.github.io/sevenbridges-r/articles/api.html">HTML</a></td>
<td align="left"><a href="https://raw.githubusercontent.com/sbg/sevenbridges-r/master/vignettes/api.Rmd">Source</a></td>
</tr>
<tr class="even">
<td align="left">Use R on the Cancer Genomics Cloud</td>
<td align="left"><a href="https://sbg.github.io/sevenbridges-r/articles/bioc-workflow.html">HTML</a></td>
<td align="left"><a href="https://raw.githubusercontent.com/sbg/sevenbridges-r/master/vignettes/bioc-workflow.Rmd">Source</a></td>
</tr>
<tr class="odd">
<td align="left">Create a Docker Container and use Command Line Interface for R</td>
<td align="left"><a href="https://sbg.github.io/sevenbridges-r/articles/docker.html">HTML</a></td>
<td align="left"><a href="https://raw.githubusercontent.com/sbg/sevenbridges-r/master/vignettes/docker.Rmd">Source</a></td>
</tr>
<tr class="even">
<td align="left">Describe and execute Common Workflow Language (CWL) Tools and Workflows in R</td>
<td align="left"><a href="https://sbg.github.io/sevenbridges-r/articles/apps.html">HTML</a></td>
<td align="left"><a href="https://raw.githubusercontent.com/sbg/sevenbridges-r/master/vignettes/apps.Rmd">Source</a></td>
</tr>
<tr class="odd">
<td align="left">
<a name="ide"></a>IDE container: Rstudio and Shiny server and more</td>
<td align="left"><a href="https://sbg.github.io/sevenbridges-r/articles/rstudio.html">HTML</a></td>
<td align="left"><a href="https://raw.githubusercontent.com/sbg/sevenbridges-r/master/vignettes/rstudio.Rmd">Source</a></td>
</tr>
<tr class="even">
<td align="left">Browse data on the Cancer Genomics Cloud via the Data Explorer, a SPARQL query, <br>or the Datasets API</td>
<td align="left"><a href="https://sbg.github.io/sevenbridges-r/articles/cgc-sparql.html">HTML</a></td>
<td align="left"><a href="https://raw.githubusercontent.com/sbg/sevenbridges-r/master/vignettes/cgc-sparql.Rmd">Source</a></td>
</tr>
</tbody>
</table>
</div>
<div id="ide-docker-image" class="section level2">
<h2 class="hasAnchor">
<a href="#ide-docker-image" class="anchor"></a>IDE Docker Image</h2>
<p>In the tutorial for <a href="#tutorials">IDE container</a> above, we built a Docker container locally from which we can launch RStudio and Shiny. To launch RStudio and Shiny Server with the Seven Bridges IDE Docker container, do the following:</p>
<pre class="shell"><code>docker run  -d -p 8787:8787 -p 3838:3838 --name rstudio_shiny_server sevenbridges/sevenbridges-r</code></pre>
<p>To mount a file system, you need to use <code>--privileged</code> with fuse.</p>
<pre class="shell"><code>docker run  --privileged -d -p 8787:8787 -p 3838:3838 --name rstudio_shiny_server sevenbridges/sevenbridges-r</code></pre>
<p>Check out the IP from Docker Machine if you are on a Mac OS.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1"></a><span class="ex">docker-machine</span> ip default</span></code></pre></div>
<p>In your browser, you can see where the RStudio server is located from the path <code>http://&lt;url&gt;:8787/</code>. For example, if 192.168.99.100 is returned, visit <code>http://192.168.99.100:8787/</code> for Rstudio.</p>
<p>For the Shiny server, each app__ is hosted at <code>http://&lt;url&gt;:3838/users/&lt;username of rstudio&gt;/&lt;app_dir&gt;</code> for the Shiny server. For example, an app called <code>01_hello</code> owned by user <code>rstudio</code> (a default user) has the path <code>http://&lt;url&gt;:3838/users/rstudio/01_hello/</code>. To develop your Shiny apps as an Rstudio user, you can login your RStudio server and create a folder in your home folder called <code>~/ShinyApps</code>. There, you can develop shiny apps in that folder. For example, you can create an app called <code>02_text</code> at <code>~/ShinyApps/02_text/</code>.</p>
<p>Log into your RStudio at <code>http://&lt;url&gt;:8787</code>. Then, try to copy an app to your home folder, as follows:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/files2">dir.create</a></span>(<span class="st">"~/ShinyApps"</span>)</span>
<span id="cb18-2"><a href="#cb18-2"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/files">file.copy</a></span>(</span>
<span id="cb18-3"><a href="#cb18-3"></a>  <span class="st">"/usr/local/lib/R/site-library/shiny/examples/01_hello/"</span>,</span>
<span id="cb18-4"><a href="#cb18-4"></a>  <span class="st">"~/ShinyApps/"</span>,</span>
<span id="cb18-5"><a href="#cb18-5"></a>  <span class="dt">recursive =</span> <span class="ot">TRUE</span></span>
<span id="cb18-6"><a href="#cb18-6"></a>)</span></code></pre></div>
<p>If you are logged in as user <code>rstudio</code>, visit <code>http://192.168.99.100:3838/rstudio/01_hello</code>. You should be able to see the “hello” example.</p>
<p><em>Note</em>: Generic Shiny apps can also be hosted at <code>http://&lt;url&gt;:3838/</code> or, for a particular app, at <code>http://&lt;url&gt;:3838/&lt;app_dir&gt;</code>. Inside the Docker container, it’s hosted under <code>/srv/shiny-server/</code>.</p>
</div>
<div id="faq" class="section level2">
<h2 class="hasAnchor">
<a href="#faq" class="anchor"></a>FAQ</h2>
<p>The best place to ask questions about the <code>sevenbridges</code> package is the <a href="https://groups.google.com/forum/#!forum/sevenbridges-r">mailing list</a>.</p>
<ul>
<li><p><strong>Q</strong>: Does this package support Seven Bridges’ API v1 which was not CWL compatible?<br><strong>A</strong>: No. This package only supports API v2 +. For API v1, please check out the <a href="https://github.com/nanxstats/sbgr">sbgr</a> package. Note that API v1 and associated legacy project types will be deprecated eventually.</p></li>
<li><p><strong>Q</strong>: Which version of the Common Workflow Language (CWL) is supported?<br><strong>A</strong>: We support draft 2 and are making progress on supporting draft 3.</p></li>
<li><p><strong>Q</strong>: Is there a Python binding for the API?<br><strong>A</strong>: Yes, the official Python client is <a href="https://github.com/sbg/sevenbridges-python">here</a>. Recipes and tutorials using the Python bindings are <a href="https://github.com/sbg/okAPI">here</a>.</p></li>
<li><p><strong>Q</strong>: Why do I get warning messages when I use the API R client?<br><strong>A</strong>: The warning only exists in Rstudio and is potentially a bug in Rstudio. To ignore, it use <code><a href="https://www.rdocumentation.org/packages/base/topics/options">options(warn = -1)</a></code></p></li>
<li><p><strong>Q</strong>: I still have problems despite dismissing the messages.<br><strong>A</strong>: Please try to use the latest package on GitHub or <a href="https://bioconductor.org/install/#update-bioconductor-packages">update installed Bioconductor packages</a>. This usually includes the most recent bug fixes.</p></li>
</ul>
</div>
<div id="events" class="section level2">
<h2 class="hasAnchor">
<a href="#events" class="anchor"></a>Events</h2>
<table class="table">
<colgroup>
<col width="24%">
<col width="35%">
<col width="40%">
</colgroup>
<thead><tr class="header">
<th align="left">Time</th>
<th align="left">Event</th>
<th align="left">Location</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Jan 12, 2017</td>
<td align="left">
<a href="https://www.meetup.com/Boston-R-Bioconductor-for-genomics/events/235580582/">Genomics in the Cloud - Boston Bioconductor Meetup</a> (talk) [<a href="https://nanx.me/talks/bioc-meetup-cgc-170112.pdf">slides</a>]</td>
<td align="left">Dana-Farber Cancer Institute, Boston, MA</td>
</tr>
<tr class="even">
<td align="left">Sep 12 - 14, 2016</td>
<td align="left">
<a href="https://www.stats.ox.ac.uk/events/probgen16">Probabilistic Modeling in Genomics</a> (poster)</td>
<td align="left">University of Oxford, Oxford, UK</td>
</tr>
<tr class="odd">
<td align="left">May 27 - 29, 2016</td>
<td align="left">
<a href="http://china-r.org/bj2016/index.html">The 9th China-R Conference</a> (talk)</td>
<td align="left">Renmin University of China, Beijng, China</td>
</tr>
<tr class="even">
<td align="left">Jun 27 - 30, 2016</td>
<td align="left">
<a href="http://user2016.org/">The R User Conference 2016</a> (talk)</td>
<td align="left">Stanford University, Stanford, CA</td>
</tr>
<tr class="odd">
<td align="left">Jun 24 - 26, 2016</td>
<td align="left">
<a href="https://bioconductor.org/help/course-materials/2016/BioC2016/">BioC 2016: Where Software and Biology Connect</a> (workshop)</td>
<td align="left">Stanford University, Stanford, CA</td>
</tr>
<tr class="even">
<td align="left">Apr 1 - 3, 2016</td>
<td align="left">
<a href="https://www.cancergenomicscloud.org/hacking-cancer/">NCI Cancer Genomics Cloud Hackathon</a> (tutorial)<br>[<a href="https://sbg.github.io/sevenbridges-r/articles/bioc-workflow.html">HTML</a>] [<a href="https://raw.githubusercontent.com/sbg/sevenbridges-r/master/vignettes/bioc-workflow.Rmd">R Markdown Source</a>]</td>
<td align="left">Seven Bridges Genomics, Inc., Boston, MA</td>
</tr>
</tbody>
</table>
</div>
<div id="contribute" class="section level2">
<h2 class="hasAnchor">
<a href="#contribute" class="anchor"></a>Contribute</h2>
<p>Please file bug reports/feature requests on the <a href="https://github.com/sbg/sevenbridges-r/issues">issue page</a>, or create pull requests <a href="https://github.com/sbg/sevenbridges-r/pulls">here</a>.</p>
<p>Contributors should read the <a href="https://drive.google.com/file/d/0B9ms5nmyMyvSY3ZmeDRqV1p4SHc/view?usp=sharing">Seven Bridges Notice to Contributors</a> and sign the <a href="https://secure.na1.echosign.com/public/esignWidget?wid=CBFCIBAA3AAABLblqZhAqt_9rHEqy2MggS0uWRmKHUN2HYi8DWNjkgg5N68iKAhRFTy7k2AOEpRHMMorxc_0*">Seven Bridges Contributor Agreement</a> before submitting a pull request.</p>
</div>
<div id="copyright" class="section level2">
<h2 class="hasAnchor">
<a href="#copyright" class="anchor"></a>Copyright</h2>
<p>© 2018 Seven Bridges Genomics, Inc. All rights reserved.</p>
<p>This project is licensed under the terms of the <a href="LICENSE">Apache License 2.0</a>.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Download from BIOC at <br><a href="https://www.bioconductor.org/packages/sevenbridges">https://​www.bioconductor.org/​packages/​sevenbridges</a>
</li>
<li>Browse source code at <br><a href="https://github.com/sbg/sevenbridges-r">https://​github.com/​sbg/​sevenbridges-r</a>
</li>
<li>Report a bug at <br><a href="https://github.com/sbg/sevenbridges-r/issues">https://​github.com/​sbg/​sevenbridges-r/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>Apache License 2.0 | file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Nan Xiao <br><small class="roles"> Author, maintainer </small>  </li>
<li>Tengfei Yin <br><small class="roles"> Author </small>  </li>
<li>Seven Bridges Genomics <br><small class="roles"> Copyright holder, funder </small>  </li>
<li><a href="authors.html">All authors...</a></li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://travis-ci.org/sbg/sevenbridges-r"><img src="https://travis-ci.org/sbg/sevenbridges-r.svg?branch=master" alt="Build Status"></a></li>
<li><a href="https://ci.appveyor.com/project/nanxstats/sevenbridges-r"><img src="https://ci.appveyor.com/api/projects/status/github/sbg/sevenbridges-r?branch=master&amp;svg=true" alt="AppVeyor Build Status"></a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/sevenbridges.html#since"><img src="https://www.bioconductor.org/shields/years-in-bioc/sevenbridges.svg" alt="BioC"></a></li>
<li><a href="https://bioconductor.org/packages/stats/bioc/sevenbridges/"><img src="https://www.bioconductor.org/shields/downloads/sevenbridges.svg" alt="Downloads"></a></li>
<li><a href="https://hub.docker.com/r/sevenbridges/sevenbridges-r/"><img src="https://img.shields.io/docker/pulls/sevenbridges/sevenbridges-r.svg" alt="Docker Pulls"></a></li>
</ul>
</div>
</div>
</div>

      <footer><div class="copyright">
  <p>Developed by Nan Xiao, Tengfei Yin, Seven Bridges Genomics.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
